server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/auth/provider/local/LocalAuthProvider.java

```diff
@@ -59,22 +70,49 @@ public String validateLocalAuth(@NotNull DBRProgressMonitor monitor,
             throw new DBException("Invalid user name or password");
         }
 
-        String storedPasswordHash = CommonUtils.toString(storedCredentials.get(CRED_PASSWORD), null);
+        String passwordSha = CommonUtils.toString(userCredentials.get(LocalAuthProviderConstants.CRED_PASSWORD), null);
+        if (CommonUtils.isNotEmpty(CommonUtils.toString(userCredentials.get(CRED_PASSWORD_MD_5), null))) {
+            validatePasswordMd5(userCredentials, storedCredentials);
+            if (securityController instanceof SMAdminController adminController) {
+                adminController.setUserCredentials(userName, authProvider.getId(), userCredentials);
+            } else {
+                throw new DBException("User password hash update is not supported in current context");
+            }
+        } else {
+            validatePasswordSha(passwordSha, storedCredentials);
+        }
+
+        return activeUserId == null ? userName : activeUserId;
+    }
+
+    private void validatePasswordSha(String passwordSha, Map<String, Object> storedCredentials) throws DBException {
+        try {
+            if (!Argon2IdHasher.verify(CommonUtils.toString(storedCredentials.get(LocalAuthProviderConstants.CRED_PASSWORD)), passwordSha)) {
```

Comment:
Let's move all game with argon inside security utils. We mustn't know what algorithms are used on this level of abstraction.